src/wp-admin/includes/class-wp-plugin-dependencies.php

```diff
@@ -0,0 +1,688 @@
+		return get_option( self::PENDING_PLUGIN_ACTIVATIONS_OPTION, array() );
+	}
+
+	/**
+	 * Set plugin to the to-be-activated queue.
+	 *
+	 * @since 5.9.0
+	 * @access private
+	 *
+	 * @param string $plugin The plugin file.
+	 *
+	 * @return void
+	 */
+	private function add_plugin_to_queue( $plugin ) {
+		$queue = $this->get_plugins_to_activate();
+		if ( in_array( $plugin, $queue, true ) ) {
+			return;
+		}
+		$queue[] = $plugin;
+
+		update_option( self::PENDING_PLUGIN_ACTIVATIONS_OPTION, $queue );
+	}
+
+	/**
+	 * Remove plugin from the to-be-activated queue.
+	 *
+	 * @since 5.9.0
+	 * @access private
+	 *
+	 * @param string $plugin The plugin file.
+	 *
+	 * @return void
+	 */
+	private function remove_plugin_from_queue( $plugin ) {
+		$queue = $this->get_plugins_to_activate();
+		if ( ! in_array( $plugin, $queue, true ) ) {
+			return;
+		}
+
+		update_option( self::PENDING_PLUGIN_ACTIVATIONS_OPTION, array_diff( $queue, array( $plugin ) ) );
+	}
+
+	/**
+	 * Check if a plugin is part of a circular dependencies loop.
+	 *
+	 * @since 5.9.0
+	 * @access private
+	 *
+	 * @param string                   $plugin_file The plugin file.
+	 * @param array<int|string, mixed> $previous    If this is a dependency of a dependency,
+	 *                                              this array contains all previous levels of dependencies.
+	 *
+	 * @return bool
+	 */
+	private function in_circular_dependency( $plugin_file, $previous = array() ) {
+		if ( isset( $this->circular_dependencies[ $plugin_file ] ) ) {
+			return $this->circular_dependencies[ $plugin_file ];
+		}
+
+		if ( in_array( $plugin_file, $previous, true ) ) {
+			$this->circular_dependencies[ $plugin_file ] = true;
+		}
+
+		$plugin_dependencies = $this->get_plugin_dependencies( $plugin_file );
+
+		foreach ( $plugin_dependencies as $dependency ) {
+			$dependency_file = $this->get_plugin_file_from_slug( $dependency['slug'] );
+			if ( $dependency_file && $this->in_circular_dependency( $dependency_file, array_merge( $previous, array( $plugin_file ) ) ) ) {
+				$this->circular_dependencies[ $plugin_file ]     = true;
+				$this->circular_dependencies[ $dependency_file ] = true;
+			}
+		}
+
+		if ( ! isset( $this->circular_dependencies[ $plugin_file ] ) ) {
+			$this->circular_dependencies[ $plugin_file ] = false;
+		}
+
+		return $this->circular_dependencies[ $plugin_file ];
+	}
+
+	/**
+	 * Get plugin file from its slug.
+	 *
+	 * @since 5.9.0
+	 * @access private
+	 *
+	 * @param string $slug The plugin slug.
+	 *
+	 * @return string|false Returns the plugin file on success, false on failure.
+	 */
+	private function get_plugin_file_from_slug( $slug ) {
+		$plugins = $this->get_plugins();
+		foreach ( array_keys( $plugins ) as $plugin ) {
+			if ( 0 === strpos( $plugin, "$slug/" ) || 0 === strpos( $plugin, "$slug\\" ) ) {
+				return $plugin;
+			}
+		}
+		return false;
+	}
+}
+
+global $plugin_dependencies;
+$plugin_dependencies = new WP_Plugin_Dependencies();
```

Comment:
I'm not sure why this is using a singleton but if it is to do so, then namespacing the global would be dandy.

```suggestion
global $wp_plugin_dependencies;
$wp_plugin_dependencies = new WP_Plugin_Dependencies();
```

:1234: Same change would be required throughout